### PR TITLE
Steph's version of dnd with ties, fixed up

### PR DESCRIPTION
**Pull request by jodawi**

### CODE REVIEW

**Review by symmetry — APPROVED** [review]
lgtm, at least afaict.  next assignment: write a set of tests of this function using the parameterized package (https://pypi.org/project/parameterized/) and also use the line coverage tool of your choice to achieve at least 90% line coverage by 11:59 Sunday night.  

prefer as close to 100% line coverage as possible but correct code + 90% coverage is a successful completion of assignment -- 100% coverage and late is NOT a successful completion.  That's about 1/2 to 1/4 the speed you'd be expected to work at to complete this task as a Python dev, fyi)

**Comment by jodawi** [comment]
Mebb later. I currently reading Java Concurrency in Practice.

File: algraph.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+            
+    def addVerts(self, verts):
+        for vert in verts:
+            if self.adjacency_list.get(vert) is None:
+                self.adjacency_list[vert] = set()
+    
+    def findTopologicalOrdering(self):
+        indegree = {}
+        for vert in self.adjacency_list.keys():
+            indegree[vert] = 0
```
Comment (symmetry) [inline algraph.py:22]: Purt sure that

`indegree = { vert:0 for vert in self.adjacency_list.keys() }`

is more pythonic

File: algraph.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
+
+class ALGraph:
+    """adjacency_list graph"""
+
+    def __init__(self):
+        self.adjacency_list = {}
```
Comment (symmetry) [inline algraph.py:6]: Where do you keep track of ALL of your vertices/nodes?  (What if a vertex/node has no edges associated with it?  e.g. person who just joined Facebook and hasn't added any friends yet.)

File: algraph.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
```
Comment (symmetry) [inline algraph.py:7]: Prefer keeping track of vertices explicitly -- if so, suggest adding functions `addVert(self, vert)` and `addEdge(self, edge)` that are called by `addVerts()` and `addEdges()` respectively, and allow the constructor to optionally take verts (as a list) and edges (as a list or tupl of 2-lists or 2-tuples), e.g.

```
class ALGraph:
    def __init__(self, verts=None, edges=None, is_directed=True):
        self.adjacency_list = {}
        self.addVerts(verts)
        self.addEdges(edges)
        self.is_directed = is_directed

    def addVert(self, vert):
        if vert is not None:
            self.adjacency_list[vert] = (self.adjacency_list[vert] is None) ? set() : self.adjacency_list[vert]

    def __addDirectedEdge(self, edge):
        self.adjacency_list[edge[0]].add(edge[1])

    def addEdge(self, edge):
        # could force list of verts to be only those passed in explicitly as vertices, in which case you'd want to check that both source and sink are legal verts
        # however, let's make the set of legal verts the union of things originally passed as verts AND those mentioned as endpoints of edges
        if edge is not None and len(edge) > 1:
            self.addVert(edge[0])
            self.addVert(edge[1])
            self.__addDirectedEdge(edge[0], edge[1])
            if not self.is_directed:
                 self.__addDirectedEdge(edge[1], edge[0])

    def addVerts(self, verts):
        if verts is not None:
            for vert in verts:
                self.addVert(vert)

    def addEdges(self, edges):
        if edges is not None:
            for edge in edges:
                self.addEdge(edge[0], edge[1])
```
I'm sure there's a typo or bug in there somewhere, but you get the idea -- anyway, given the above, rewrite findTopologicalOrdering() to work correctly?

File: algraph.py (hunk 2 above)
Comment (jodawi) [reply algraph.py:6]: It's just a toy graph to allow for implementing the findTopologicalOrdering algorithm.

**Pull request merged**